internal/encryption/encryptor.go

```diff
@@ -69,136 +74,266 @@ func (d *DocEncryptor) SetStore(store datastore.DSReaderWriter) {
+
+	return cipherText, err
 }
 
 // Decrypt decrypts the given cipherText that is associated with the given docID and fieldName.
 // If the corresponding encryption key is not found, it returns nil.
-func (d *DocEncryptor) Decrypt(docID, fieldName string, cipherText []byte) ([]byte, error) {
-	encKey, err := d.fetchEncryptionKey(docID, fieldName)
+func (d *DocEncryptor) Decrypt(
+	encStoreKey core.EncStoreDocKey,
+	cipherText []byte,
+) ([]byte, error) {
+	if d.store == nil {
+		return nil, ErrNoStorageProvided
+	}
+	encKey, err := d.fetchByEncStoreKey(encStoreKey)
 	if err != nil {
 		return nil, err
 	}
 	if len(encKey) == 0 {
 		return nil, nil
 	}
-	return DecryptAES(cipherText, encKey)
+	return crypto.DecryptAES(nil, cipherText, encKey, nil)
 }
 
-// fetchEncryptionKey fetches the encryption key for the given docID and fieldName.
-// If the key is not found, it returns an empty key.
-func (d *DocEncryptor) fetchEncryptionKey(docID string, fieldName string) ([]byte, error) {
-	if d.store == nil {
-		return nil, ErrNoStorageProvided
+func (d *DocEncryptor) fetchByEncStoreKey(storeKey core.EncStoreDocKey) ([]byte, error) {
+	if encryptionKey, ok := d.cache[storeKey]; ok {
+		return encryptionKey, nil
 	}
-	// first we try to find field-level key
-	storeKey := core.NewEncStoreDocKey(docID, fieldName)
 	encryptionKey, err := d.store.Get(d.ctx, storeKey.ToDS())
 	isNotFound := errors.Is(err, ds.ErrNotFound)
 	if err != nil {
-		if !isNotFound {
-			return nil, err
-		}
-		// if previous fetch was for doc-level, there is nothing else to look for
-		if fieldName == "" {
-			return nil, nil
-		}
-		if shouldEncryptIndividualField(d.conf, fieldName) {
+		if isNotFound {
 			return nil, nil
 		}
-		// try to find doc-level key
-		storeKey.FieldName = ""
-		encryptionKey, err = d.store.Get(d.ctx, storeKey.ToDS())
-		isNotFound = errors.Is(err, ds.ErrNotFound)
-		if err != nil && !isNotFound {
-			return nil, err
-		}
+		return nil, err
 	}
+
+	d.cache[storeKey] = encryptionKey
 	return encryptionKey, nil
 }
 
-// EncryptDoc encrypts the given plainText that is associated with the given docID and fieldName with
+func (d *DocEncryptor) storeByEncStoreKey(storeKey core.EncStoreDocKey, encryptionKey []byte) error {
+	d.cache[storeKey] = encryptionKey
+	return d.store.Put(d.ctx, storeKey.ToDS(), encryptionKey)
+}
+
+// GetKey returns the encryption key for the given docID, (optional) fieldName and block height.
+func (d *DocEncryptor) GetKey(encStoreKey core.EncStoreDocKey) ([]byte, error) {
+	if d.store == nil {
+		return nil, ErrNoStorageProvided
+	}
+	encryptionKey, err := d.fetchByEncStoreKey(encStoreKey)
+	if err != nil {
+		return nil, err
+	}
+	return encryptionKey, nil
+}
+
+// getGeneratedKeyFor returns the generated key for the given docID and fieldName.
+func (d *DocEncryptor) getGeneratedKeyFor(
+	docID string,
+	fieldName immutable.Option[string],
+) (immutable.Option[core.EncStoreDocKey], []byte) {
+	for _, key := range d.generatedKeys {
```

Comment:
todo: You don't need to keep this list of generated keys to be able to resolve this function. Unless specifically told otherwise, the DB should continue to use the key from the last block within the document. If this is the first block (create document) then we need to generate a new key.

You can get the last generated key from the current head of the document in the headstore.